### PR TITLE
fix: websocket closed (tentative)

### LINKED ISSUE #28523: Error: WebSocket connection closed (in cypress 13.6)
### Current behavior

Cypress reports `Error: WebSocket connection closed` and then stops running tests and `cypress.run` returns with
```json
{
  "failures": 1,
  "message": "Could not find Cypress test run results"
}
```

More context:
* This happens sporadically, in the same spot, for a few projects. Maybe ~10% of the time for those projects.
* This began after upgrading from Cypress 12.17 -> 13.6.1. Before/after volume of this error message:
image

### Desired behavior

Don't throw errors. Or give more information as to why the errors are happening.

### Test code to reproduce

not sure how to reproduce; happens sporadically but consistently 

### Cypress Version

13.6.1

### Node version

18.16.1

### Operating System

linux/docker (CI/CD)

### Debug Logs

_No response_

### Other

_No response_

### PR DESCRIPTION
**Pull request by cacieprins**
- Closes https://github.com/cypress-io/cypress/issues/28523

### Additional details

*This is a tentative fix*. This issue has been very difficult if not impossible to reproduce, and this is my *best guess* solution. 

Occasionally, certain systems can send CDP commands that will fail due to the CRI websocket being closed with those commands mid-flight. This is most common with video recording ACK commands, but can potentially occur with any automation command. Prior to `chrome-remote-interface@0.33.0`, it seems as though these commands failed silently. `0.33.0` introduced behavior where it will throw an error in this situation. This exception was uncaught, causing Cypress to intermittently exit with code 1 even thought a given spec had succeeded.

This PR:

* Adds additional verbose debug steps to the CRI client, for non-send/receive websocket events
* handles `Error: WebSocket connection closed` errors similarly to `WebSocket is not open` and `WebSocket already in CLOSING or CLOSED state` errors: the command is enqueued, and reconnection to the CRI websocket is attempted
* When reconnecting to the CRI websocket due to a failed command, CRI Client now checks to see if there is an in-flight reconnection process and if so, awaits that instead of kicking off a new one
* removes enablement commands (e.g., `Runtime.addBinding`, `DOM.enable`) from the command queue, if present, when sending them upon reconnection. This prevents them from being re-sent when iterating through the queued commands.
* Updates integration tests so they will fail if this exception is uncaught

### Steps to test

Run server unit and integration tests. A reproduction of the original issue is unavailable, and confirmation on a fix will be pending user feedback.

### PR Tasks

- [X] Have tests been added/updated?
- [NA] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [NA] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

File: cli/CHANGELOG.md [hunk 1]
```diff
@@ -1,4 +1,12 @@
 <!-- See the ../guides/writing-the-cypress-changelog.md for details on writing the changelog. -->
+## 13.8.1
+
+_Released 4/23/2024 (PENDING)_
+
+**Bufixes:**
```
Comment (jennifer-shehane) [inline cli/CHANGELOG.md]:
```suggestion
**Bugfixes:**
```

File: cli/CHANGELOG.md [hunk 2]
```diff
@@ -1,4 +1,12 @@
 <!-- See the ../guides/writing-the-cypress-changelog.md for details on writing the changelog. -->
+## 13.8.1
+
+_Released 4/23/2024 (PENDING)_
+
+**Bugfixes:**
+
+- Fixes a bug introduced in [`13.6.0`](https://docs.cypress.io/guides/references/changelog#13-6-0) where Cypress would occasionally exit with status code 1 even when the test run was successful, due to an unhandled WebSocket exception (`Error: WebSocket connection closed`). Addresses [#28523](https://github.com/cypress-io/cypress/issues/28523).
```
Comment (AtofStryker) [inline cli/CHANGELOG.md]:
```suggestion
- Fixes a bug introduced in [`13.6.0`](https://docs.cypress.io/guides/references/changelog#13-6-0) where Cypress would occasionally exit with status code 1, even when a test run was successfully, due to an unhandled WebSocket exception (`Error: WebSocket connection closed`). Addresses [#28523](https://github.com/cypress-io/cypress/issues/28523).
```

**Review by AtofStryker — APPROVED** [review]

File: packages/server/lib/browsers/cri-client.ts [hunk 3]
```diff
@@ -11,12 +11,15 @@ import type { SendDebuggerCommand, OnFn, CdpCommand, CdpEvent } from './cdp_auto
...
-const debugVerboseSend = debugModule('cypress-verbose:server:browsers:cri-client:send:[-->]')
+const debugVerboseSend = debugModule(`${debugVerbose.namespace}:send:[-->]`)
 // debug using cypress-verbose:server:browsers:cri-client:recv:*
-const debugVerboseReceive = debugModule('cypress-verbose:server:browsers:cri-client:recv:[<--]')
+const debugVerboseReceive = debugModule(`${debugVerbose.namespace}:recv:[<--]`)
+// debug using cypress-verbose:server:browsers:cri-client:err:*
+const debugVerboseLifecycle = debugModule(`${debugVerbose.namespace}:ws`)
 
-const WEBSOCKET_NOT_OPEN_RE = /^WebSocket is (?:not open|already in CLOSING or CLOSED state)/
+const WEBSOCKET_NOT_OPEN_RE = /^WebSocket (?:connection closed|is (?:not open|already in CLOSING or CLOSED state))/
```
Comment (ryanthemanuel) [inline packages/server/lib/browsers/cri-client.ts:30]: Can we document what the various messages are that we're covering here? I always like to have examples around regexes for my future self.

**Review by ryanthemanuel — APPROVED** [review]

File: packages/server/lib/browsers/cri-client.ts (hunk 3 above)
Comment (cacieprins) [reply packages/server/lib/browsers/cri-client.ts:30]: Addressed in https://github.com/cypress-io/cypress/pull/29347/commits/a613bca09ada44aaf708e6e52cd8507e04272678

**Review by ryanthemanuel — APPROVED** [review]

**Pull request merged**